rfd/9999-passwordless-fido2.md

```diff
@@ -0,0 +1,551 @@
+            │                   │                          │                                          │
+            │                   │         *chooses*        │                                          │
+            │                   │ ─────────────────────────>                                          │
+            │                   │                          │                                          │
+            │               sign challenge                 │                                          │
+            │<─────────────────────────────────────────────│                                          │
+            │                   │                          │                                          │
+            │        user handle, signed challenge         │                                          │
+            │─────────────────────────────────────────────>│                                          │
+            │                   │                          │                                          │
+            │                   │                          │ Authenticate(userHandle, signedChallenge)│
+            │                   │                          │ ─────────────────────────────────────────>
+            │                   │                          │                                          │
+            │                   │                          │                    ok                    │
+            │                   │                          │ <─────────────────────────────────────────
+     ┌──────┴──────┐           user                    ┌───┴───┐                                  ┌───┴────┐
+     │authenticator│            ┌─┐                    │browser│                                  │Teleport│
+     └─────────────┘            ║"│                    └───────┘                                  └────────┘
+                                └┬┘
+                                ┌┼┐
+                                 │
+                                ┌┴┐
+```
+
+(Simplified Web authentication diagram.)
+
+All user<->browser interaction steps depend on the browser implementation. The
+observations below are meant to inform RFD readers on the status quo at the time
+of writing.
+
+(1) may be skipped if there is only a single authenticator present in the
+machine. In case of ambiguity (more than one authenticator), then the user is
+prompted to pick their authenticator of choice by tapping it.
+
+(2) and (3) may be skipped for biometric authenticators - the initial tap is
+sufficient for both verification and presence checks.
+
+(4) may also be skipped if there is a single resident key for the RPID (Relying
+Party ID) in the authenticator, otherwise a menu with options shows up.
+
+### CLI clients (aka `tsh`)
+
+There are essentially two options for CLI clients / `tsh`:
+
+1. Leverage browsers and delegate FIDO2 authentication to them; or
+2. Write CLI-native FIDO authentication.
```

Comment:
I do like the option (1) with all the benefits it provides. As mentioned, it gives us a free-multi-platform support which should not be underestimated. 

As for downside, I am not sure if purely FIDO2 CLI-based flow matters that much. This is because `tsh` will always require a browser for SSO which is by far the most popular Teleport auth preference. And since we know that browser UX _wins_ over CLI (with less taps), it seems like option (1) is preferable.

And another benefit is the Teleterm (upcoming native Teleport client) will be able to use `tsh` APIs as is :)

Reply:
Yep, I agree with you there. It's hard to ignore the advantages of this option.

Reply:
@codingllama @alex-kovoy what if we limit passwordless flow only if folks are using teleterm? Can this credentials handler used in teleport terminal?

Reply:
We can support both options (1) and (2) in `teleterm`. It's just option (2) will require minor changes (when prompting for user input) which is not a big deal.

Reply:
While delegating FIDO2 authentication to browsers is attractive, won’t this “train” users to trust a browser pop-up (a thing attackers can also easily create) when starting the FIDO ceremony? Restricting this to Teleterm would be best since you would also have greater control in respect to using the default user agent for the system (through Electron it's also possible to enforce certificate pinning).

Reply:
I think this ship has sailed with the various Cloud CLI tools doing browser popups, but it's a fair point.